### PR TITLE
[optimization] Slice(constant1)->constant2 optimization

### PR DESCRIPTION
**Pull request by SumedhArani**
*Description*: Compile time graph optimization. In reference with #2168 
*Testing*: Need to write graphs to test the optimization exhaustively

### CODE REVIEW

**Comment by opti-mix** [comment]
@SumedhArani Thanks for working on this!

File: lib/Optimizer/GraphOptimizer.cpp [hunk 1]
```diff
@@ -1727,6 +1727,22 @@ static void optimizeSliceOfSplat(Function *F) {
...
+/// Slice(constant1) -> constant2
+static void optimizeSliceOfConstant(Function *F) {
+  for (auto &node : F->getNodes()) {
+    auto *sliceNode = dyn_cast<SliceNode>(&node);
+    if (!sliceNode)
+      continue;
+    auto *constantNode = dyn_cast<Constant>(sliceNode->getInput());
+    if (!constantNode)
+      continue;
+    auto *newConstantNode = F->getParent()->createConstant(
```
Comment (opti-mix) [inline lib/Optimizer/GraphOptimizer.cpp]: What happens with the payload of this new constant node? It needs to be backed by a tensor that represents the slice of the original constant, or?

But I don't see here how you initialize this payload.

**Comment by jfix71** [comment]
@SumedhArani Looks like a good start! Can you add a test to `GraphOptzTest` that checks that this optimization occurs correctly?

**Comment by beicy** [comment]
Please run `./utils/format.sh fix ` after your change. This command will handle the format issue :)

File: lib/Optimizer/GraphOptimizer.cpp (hunk 1 above)
Comment (SumedhArani) [reply lib/Optimizer/GraphOptimizer.cpp]: I missed that. Thank you for pointing that out. I will make the necessary changes and update.

**Comment by rdzhabarov** [comment]
@SumedhArani do you plan to work on this PR?

**Comment by SumedhArani** [comment]
@rdzhabarov I am genuinely interested to get this work done at the earliest. However, I was bogged done with prior commitments. I should have an update for you on this by the end of this week. 

I really apologize for the delay.

**Comment by SumedhArani** [comment]
I know I have still not made any changes. I really appreciate your patience. The internship search has consumed a lot of time for me. I'll have made the required changes by the end of this week, 10 March.

Thank you once again for your patience.

File: lib/Optimizer/GraphOptimizer.cpp (hunk 1 above)
Comment (SumedhArani) [reply lib/Optimizer/GraphOptimizer.cpp]: I have spent the last couple of days reading existing code and through the documentation. 
I have been however stuck at the point of initializing the payload of the newConstantNode to be the sliced version of the original constant.
Could you give me pointers regarding finding the sliced payload of a tensor?

Comment (opti-mix) [reply lib/Optimizer/GraphOptimizer.cpp]: @SumedhArani `Constant` has a `getPayload` method returning a tensor with a payload of this constant. Now you need to extract a proper slice from it. And `Tensor` has a number of methods for manipulating its content, e.g. `copySlice`, `copyConsecutiveSlices`, etc. You should try to use those method to form a new Tensor representing the sliced version of the original constant. Once you have, you can use `createConstant` that takes a tensor as one of the parameters or you can create a `Constant` using type, shape, etc and then set its payload by means of calling the `Constant::assign` method.

Comment (SumedhArani) [reply lib/Optimizer/GraphOptimizer.cpp]: Thanks @opti-mix !

It's of much help. Thanks for taking the time to respond back to me! 👍

**Pull request closed without merging**